### ISSUE TITLE
Skill suggestion - Logic issues & GS-Skills

### ISSUE THREAD

**Issue by alexseyk** [ToranSharma/Duo-Strength#149]
1. Grammar Skills are offered... i thought you fixed it as per "Suggested skill on for tree level 2 trees not to suggest grammar skills as they will not contribute towards getting to tree level 3."  Is it because my tree is Level 1 ? in any case L2 GS should not be offered, since (so far) there is no L3 for GS, as you know.  
2. Show / Hide with Broken & Strengthen Logic:  It seem that with Hide (On, Off) or Hide (Off, On)  = Hide (Off) for some reason I was expecting it to be Hide (On). I.e. user selected option to be a priority over the default. By my logic such behavior would be perfectly fine if you rename "Hide" to  "Show" and set default to On.
<img width="1017" alt="Screen Shot 2021-03-02 at 13 00 09" src="https://user-images.githubusercontent.com/60418150/109605231-6615ad80-7b57-11eb-894d-2d8d37fbb99c.png">

**Comment by ToranSharma** [comment]
1) Yeah as you said, I did make it so that L2 grammar skills shouldn't but suggested as they cannot get to L3. I have found the issue, I was filtering out the grammar skills from the skills that are being selected from, but didn't replace the list with the filtered version.

``` javascript
 // Remove crown 2 grammar skills as these are at max level,
 // so practising them will not contribute to getting to tree level 3.
 skillsByCrowns[2].filter(skill => skill.category !== "grammar");
```
Should have been
``` javascript
skillsByCrowns[2] = skillsByCrowns[2].filter(skill => skill.category !== "grammar");
```

Fixed in d1831c1

**Comment by ToranSharma** [comment]
Regarding 2:

I can understand how it is not as you might initially expect. The confusing phrasing of these options is due to the design and philosophy of the old options page.

The way I had made the options page initially was so that all the options would be checked by default. There "enable all" and "disable all" buttons at the bottom of the options page. The idea was that enable all would be the same as the default settings (all items ticked), and disable all would be everything off (all items unticked).

However I also had (and still have) the belief that the only options that should be enabled by default are those that add information. Any that hide or obscure things, should be off by default. These two things together meant that some options needed to be phrased in the negative, so that something wouldn't happen by default, but the option for it would be ticked.

With the options page redesign I have abandoned the desire to have the default options as all ticked, I have have a separate file with the default values in now, and another file with the values for the display of all the options to be disabled.

I am still hampered by the old options as I don-'t want to break anyone's settings, so changing the options needs to be done carefully, and only cosmetically, the underlying default cannot be changed.

I will see if I can change the wording and maybe invert the displayed checkbox to make it more intuitive. De Morgan's laws may cause me some issues here.

**Comment by alexseyk** [comment]
Well, I failed logic in university, may be that's why i am so confused with  negative phrasing and  conjunctions and disjunctions :) So just ignore my ramblings

**Comment by ToranSharma** [comment]
:smile: I haven't made it the most clear. The behaviour is correct so I'll close the issue and see if I can come up with some way of reworking it so that it is a bit less mind bending.

**Comment by alexseyk** [comment]
Hi
I think something went wrong again ...  
I believe that Grammar Skills with Crown Level <=1 should be suggested ahead of regular L2 skills 

I have some GS L1 skills above normal L2 skills and they are not suggested for study

**Comment by ToranSharma** [comment]
This should only be true on L0 trees as skills are locked. Onese a user is
at ≥L1on all skills, the skill suggestions don't need to be in any
particular order as all skills are unlocked. There is a setting in the
skill suggestion section for the selection process. By default it is random
to tomry to emulate some vague form of spaced repetition, getting the all
the skills up to the next crown level at about the same rate. However you
can change that to first or last to do then in tree or reverse tree order.

**Comment by alexseyk** [comment]
I agree that it should... but it is not the case, hence i am reporting the bug to you. I have already changed from "random" to "first" and all skills are >=1. See screenshots for proof . 

"Activ. 4" is suggested , while it should be "Past Tense 2" 

<img width="867" alt="Screen Shot 2021-03-04 at 17 43 48" src="https://user-images.githubusercontent.com/60418150/109952770-cc463000-7d11-11eb-8887-8a13f9328560.png">
<img width="404" alt="Screen Shot 2021-03-04 at 17 44 49" src="https://user-images.githubusercontent.com/60418150/109952781-cfd9b700-7d11-11eb-9d07-620602a19346.png">

**Comment by ToranSharma** [comment]
Ah sorry I misunderstood. I need to look at how the tree level is calculated, I could have removed grammar skills from the consideration all together,  which is why the crowd popup is saying your tree is level 2. Which as you have a L1 skill it shouldn't be

**Comment by ToranSharma** [comment]
Activ. 4 is being suggested as it is the first L2 skill

**Comment by ToranSharma** [comment]
Something is going on so I will reopen the issue

**Comment by alexseyk** [comment]
> Activ. 4 is being suggested as it is the first L2 skill

Which it should not since there are L1 skills...  grammar skills :)

**Comment by ToranSharma** [comment]
Just looking at the tree level code now and I am removing all grammar skills before doing the calculation. This means that as shown your tree will be thought of as L2 despite the L1 grammar skills. The skill suggestion algorithm selects from skills at the current tree level which will be 2, hence the L1 grammar skills being ignored.

I will add back the grammar skills to the tree level calculation, just removing them from the number of skills at L2 so that all trees are not stuck at L2.

**Comment by ToranSharma** [comment]
Fixed in aca212f.

L0 and L1 grammar skills will be included in the tree level calculation but, L2 will not.

**Comment by alexseyk** [comment]
thanks
